lib/winapi/findfile.c

```diff
@@ -0,0 +1,129 @@
+#include <string.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <winbase.h>
+#include <xboxkrnl/xboxkrnl.h>
+#include <hal/fileio.h>
+#include <hal/winerror.h>
+#include <fileapi.h>
+
+struct FileInfo {
+    FILE_DIRECTORY_INFORMATION dirInfo;
+    char filename[MAX_PATH];
+};
+
+static void dirtofind (FILE_DIRECTORY_INFORMATION *dirInfo, LPWIN32_FIND_DATAA lpFindFileData)
+{
+    lpFindFileData->dwFileAttributes = dirInfo->FileAttributes;
+    lpFindFileData->ftCreationTime.dwLowDateTime = dirInfo->CreationTime.LowPart;
+    lpFindFileData->ftCreationTime.dwHighDateTime = dirInfo->CreationTime.HighPart;
+    lpFindFileData->ftLastAccessTime.dwLowDateTime  = dirInfo->LastAccessTime.LowPart;
+    lpFindFileData->ftLastAccessTime.dwHighDateTime = dirInfo->LastAccessTime.HighPart;
+    lpFindFileData->ftLastWriteTime.dwLowDateTime  = dirInfo->LastWriteTime.LowPart;
+    lpFindFileData->ftLastWriteTime.dwHighDateTime = dirInfo->LastWriteTime.HighPart;
+    lpFindFileData->nFileSizeHigh = dirInfo->EndOfFile.HighPart;
+    lpFindFileData->nFileSizeLow = dirInfo->EndOfFile.LowPart;
+    memcpy(lpFindFileData->cFileName, dirInfo->FileName, dirInfo->FileNameLength);
+    lpFindFileData->cFileName[dirInfo->FileNameLength] = '\0';
+    lpFindFileData->cAlternateFileName[0] = '\0';
+}
+
+HANDLE FindFirstFileA (LPCSTR lpFileName, LPWIN32_FIND_DATAA lpFindFileData)
+{
+    NTSTATUS status;
+    ANSI_STRING dirPath;
+    ANSI_STRING mask;
+    IO_STATUS_BLOCK ioStatusBlock;
+    OBJECT_ATTRIBUTES attributes;
+    struct FileInfo fileInformation;
+    HANDLE handle;
+    size_t maskOffset;
+    char fullPath[MAX_PATH+29];
+
+    assert(strlen(lpFileName) <= MAX_PATH);
+
+    status = XConvertDOSFilenameToXBOX(lpFileName, fullPath);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(status);
+        return INVALID_HANDLE_VALUE;
+    }
+
+    RtlInitAnsiString(&dirPath, fullPath);
+
+    for (maskOffset = dirPath.Length; maskOffset > 0; maskOffset--) {
+        if (dirPath.Buffer[maskOffset - 1] == '\\')
+            break;
+    }
+
+    mask.Buffer = dirPath.Buffer + maskOffset;
+    mask.Length = dirPath.Length - maskOffset;
+    mask.MaximumLength = mask.Length;
+
+    dirPath.Length = maskOffset;
+    dirPath.MaximumLength = dirPath.Length;
+
+    if (dirPath.Length == 0 || mask.Length == 0) {
+        SetLastError(ERROR_INVALID_PARAMETER);
+        return INVALID_HANDLE_VALUE;
+    }
+
+    // Ensure that "*.*" lists all files, just like on Windows
+    if (mask.Length == 3 && strncmp(mask.Buffer, "*.*", 3)) {
+        mask.Length = 0;
+    }
```

Comment:
In comments below there is discussion about `*.*` vs `*`, but I don't see the latter case handled here. Should this be handled as well?

Reply:
The code implies that "*" as mask is handled the same way on Windows and Xbox.
This bit of code only compensates for the differences.

It should still be tested though. Also see https://github.com/XboxDev/nxdk/pull/112/files/f44a0316dcc43a4d43b282d440ea86ffe394c9fd#r284989562

Reply:
How does the code compensate for the difference? It looks like if it matches `*.*` it removes the mask entirely, while `*` will make it through? Should `mask.Length` be set to 1, instead of 0, so it matches `*` instead?

Reply:
This is how XAPI also seems to handle it.

I assume it could also be set to `mask.Length = 1`, but I think setting it to zero better highlights this special case.